fasm/parse_fasm.py

```diff
@@ -0,0 +1,217 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2020  The SymbiFlow Authors.
+#
+# Use of this source code is governed by a ISC-style
+# license that can be found in the LICENSE file or at
+# https://opensource.org/licenses/ISC
+#
+# SPDX-License-Identifier: ISC
+
+from ctypes import CDLL, POINTER, CFUNCTYPE, c_char
+import os
+from sys import byteorder
+from fasm.model import \
+    SetFasmFeature, Annotation, FasmLine, tag_to_value_format
+import fasm.tags as tags
+import platform
+from pathlib import Path
+
+if platform.system() == 'Darwin':
+    parse_fasm_lib = "libparse_fasm.dylib"
+else:
+    parse_fasm_lib = "libparse_fasm.so"
+
+here = Path(os.path.dirname(os.path.realpath(__file__)))
+parse_fasm = CDLL(str(here / parse_fasm_lib))
+parse_fasm.from_string.restype = POINTER(c_char)
+parse_fasm.from_file.restype = POINTER(c_char)
+
+# The following functions decode the binary format generated by parse_fasm.
+# This is a lightweight binary format designed to be fast to decode.
+
+
+def get_header(tags, data, i):
+    """ Match a tag and retrieve length from the header. """
+    if not data[i] in tags:
+        return None, i
+
+    return int.from_bytes(data[i + 1:i + 5], byteorder), i + 5
+
+
+def tagged_string_from_bytes(tag, data, i):
+    """ Decode a tagged string. """
+    length, i = get_header(tag, data, i)
+    if length is None:
+        return None, i
+
+    return data[i:i + length].decode('ascii'), i + length
+
+
+def fasm_address_from_bytes(data, i):
+    """ Decode a FASM address: [x:y] """
+    length, i = get_header(tags.address, data, i)
+    if not length:
+        return None, None, i
+
+    assert length == 4 or length == 8
```

Comment:
It is handy to add a little data on assertion failures, like:
```
assert length == 4 or length == 8, length
```

Reply:
Done.